docs/reconfigurator.adoc

```diff
@@ -0,0 +1,183 @@
+
+Even when these things happen, we want that the system:
+
+* never moves backwards (i.e., towards a previous target)
+* converges towards the current target
+
+This is easier than it sounds.  Take the example of managing Omicron zones.
+
+[sidebar]
+.Example: managing Omicron zones
+--
+Reconfigurator manages the set of Omicron zones running on each sled.  How can we ensure that when changes are made, the system only moves forward even when there are multiple Nexus instances executing blueprints concurrently and some might be executing older versions?
+
+First, we apply a generation number to "the set of Omicron zones" on each sled.  Blueprints store _for each sled_ (1) the set of zones on that sled (and their configuration) and (2) the generation number.  Any time we want to change the set of zones on a sled, we make a new blueprint with the updated set of zones and the next generation number.  Execution is really simple: we make an API call to sled agent specifying the new set of zones _and_ the new generation number.  Sled Agent keeps track of the last generation number that it saw and rejects requests with an older one.  Now, if multiple Nexus instances execute the latest target, all will succeed and the first one that reaches each Sled Agent will actually update the zones on that sled.  If there's also a Nexus executing an older blueprint, it will be rejected.
+
+// TODO mermaid diagram showing concurrent execution
+--
+
+This approach can be applied to many other areas like DNS configuration, too.  Other areas (e.g., the process of updating database state to reflect internal IP allocations) sometimes require different, _ad hoc_ mechanisms.  In all cases, though, the goals are what we said above: attempting to execute a stale blueprint must never move the system backwards and as long as _something_ is executing the newer blueprint, the system should eventually get to the new target.
+
+== Reconfigurator control: autonomous vs. manual
+
+=== Long-term goal: autonomous operation
+
+The long-term goal is to enable autonomous operation of both the **planner** and **executor**:
+
+[source,text]
+----
+The Planner
+
+    fleet policy  (database state, inventory)   (latest blueprint)
+             \               |               /
+              \              |              /
+               +----------+  |  +----------/
+                          |  |  |
+                          v  v  v
+
+                         "planner"
+              (eventually a background task)
+                             |
+                             v                      no
+                    is a new blueprint necessary? ------> done
+                             |
+                             | yes
+                             v
+                    generate a new blueprint
+                             |
+                             |
+                             v
+                    commit blueprint to database
```

Comment:
I believe "commit to database" and "make the target" are two separate steps - maybe worth an extra node in the graph for "(attempt to) make it the current target"

Reply:
Done in 4bed612b900bb556e8cd3ec97d2394ea8d8f0ec5.